python_modules/dagster/dagster/_core/storage/partition_status_cache.py

```diff
@@ -0,0 +1,273 @@
+        partitions_def_id (Optional(str)): The serializable unique identifier for the partitions
+            definition. When this value differs from the new partitions definition, this cache
+            value needs to be recalculated. None if the asset is unpartitioned.
+        materialized_partition_subsets (Optional(str)): The serializable representation of the
+            materialized partition subsets, up to the latest storage id. None if the asset is
+            unpartitioned.
+    """
+
+    def __new__(
+        cls,
+        latest_storage_id: int,
+        partitions_def_id: Optional[str] = None,
+        materialized_partition_subsets: Optional[str] = None,
+    ):
+        check.int_param(latest_storage_id, "latest_storage_id")
+        check.opt_inst_param(
+            materialized_partition_subsets,
+            "materialized_partition_subsets",
+            str,
+        )
+        check.opt_str_param(partitions_def_id, "partitions_def_id")
+        return super(AssetStatusCacheValue, cls).__new__(
+            cls, latest_storage_id, partitions_def_id, materialized_partition_subsets
+        )
+
+    @staticmethod
+    def from_db_string(db_string):
+        if not db_string:
+            return None
+
+        cached_data = deserialize_json_to_dagster_namedtuple(db_string)
+        if not isinstance(cached_data, AssetStatusCacheValue):
+            return None
+
+        return cached_data
+
+    def deserialize_materialized_partition_subsets(
+        self, partitions_def: PartitionsDefinition
+    ) -> PartitionsSubset:
+        if not self.materialized_partition_subsets:
+            return partitions_def.empty_subset()
+
+        return partitions_def.deserialize_subset(self.materialized_partition_subsets)
+
+
+def get_materialized_multipartitions_from_tags(
+    instance: DagsterInstance, asset_key: AssetKey, partitions_def: MultiPartitionsDefinition
+) -> List[MultiPartitionKey]:
+    dimension_names = partitions_def.partition_dimension_names
+    materialized_keys: List[MultiPartitionKey] = []
+    for event_tags in instance.get_event_tags_for_asset(asset_key):
+        event_partition_keys_by_dimension = {
+            get_dimension_from_partition_tag(key): value
+            for key, value in event_tags.items()
+            if key.startswith(MULTIDIMENSIONAL_PARTITION_PREFIX)
+        }
+
+        if all(
+            [
+                dimension_name in event_partition_keys_by_dimension.keys()
+                for dimension_name in dimension_names
+            ]
+        ):
+            materialized_keys.append(
+                MultiPartitionKey(
+                    {
+                        dimension_names[0]: event_partition_keys_by_dimension[dimension_names[0]],
+                        dimension_names[1]: event_partition_keys_by_dimension[dimension_names[1]],
+                    }
+                )
+            )
+    return materialized_keys
+
+
+def _rebuild_status_cache(
```

Comment:
Nitpick: this won't always be a _re_-build, right? I.e. we'll call this the first time? If so, maybe a name like "build from scratch"?